tests/tls_handler_test.c

```diff
@@ -843,3 +843,320 @@ static int s_tls_client_channel_negotiation_success_fn(struct aws_allocator *all
+
+    /* shut down */
+    aws_channel_shutdown(incoming_args.channel, AWS_OP_SUCCESS);
+    ASSERT_SUCCESS(aws_condition_variable_wait_pred(
+        &condition_variable, &mutex, s_tls_channel_shutdown_predicate, &incoming_args));
+
+    /*no shutdown on the client necessary here (it should have been triggered by shutting down the other side). just
+     * wait for the event to fire. */
+    ASSERT_SUCCESS(aws_condition_variable_wait_pred(
+        &condition_variable, &mutex, s_tls_channel_shutdown_predicate, &outgoing_args));
+
+    /* connect again! */
+    outgoing_args.tls_negotiated = false;
+    outgoing_args.shutdown_finished = false;
+    incoming_args.tls_negotiated = false;
+    incoming_args.shutdown_finished = false;
+    ASSERT_SUCCESS(aws_client_bootstrap_new_tls_socket_channel(
+        client_bootstrap,
+        endpoint.address,
+        0,
+        &options,
+        &tls_client_conn_options,
+        s_tls_handler_test_client_setup_callback,
+        s_tls_handler_test_client_shutdown_callback,
+        &outgoing_args));
+    /* put this here to verify ownership semantics are correct. This should NOT cause a segfault. If it does, ya
+     * done messed up. */
+    aws_tls_connection_options_clean_up(&tls_client_conn_options);
+    /* wait for both ends to setup */
+    ASSERT_SUCCESS(
+        aws_condition_variable_wait_pred(&condition_variable, &mutex, s_tls_channel_setup_predicate, &incoming_args));
+
+    /* shut down */
+    aws_channel_shutdown(incoming_args.channel, AWS_OP_SUCCESS);
+    ASSERT_SUCCESS(aws_condition_variable_wait_pred(
+        &condition_variable, &mutex, s_tls_channel_shutdown_predicate, &incoming_args));
+
+    /*no shutdown on the client necessary here (it should have been triggered by shutting down the other side). just
+     * wait for the event to fire. */
+    ASSERT_SUCCESS(aws_condition_variable_wait_pred(
+        &condition_variable, &mutex, s_tls_channel_shutdown_predicate, &outgoing_args));
+
+    /* clean up */
+    aws_client_bootstrap_release(client_bootstrap);
+    ASSERT_SUCCESS(aws_server_bootstrap_destroy_socket_listener(server_bootstrap, listener));
+    aws_server_bootstrap_release(server_bootstrap);
+    aws_tls_ctx_options_clean_up(&client_ctx_options);
+    aws_tls_connection_options_clean_up(&tls_client_conn_options);
+    aws_tls_ctx_options_clean_up(&server_ctx_options);
+    aws_tls_connection_options_clean_up(&tls_server_conn_options);
+    aws_tls_ctx_destroy(client_ctx);
+    aws_tls_ctx_destroy(server_ctx);
+    aws_host_resolver_clean_up(&resolver);
+    aws_event_loop_group_clean_up(&el_group);
+    aws_tls_clean_up_static_state();
+    return AWS_OP_SUCCESS;
+}
+AWS_TEST_CASE(tls_server_multiple_connections, s_tls_server_multiple_connections_fn)
+
+struct shutdown_listener_tester {
+    struct aws_socket *listener;
+    struct aws_server_bootstrap *server_bootstrap;
+    struct tls_test_args *outgoing_args;
+    struct aws_socket *socket;
```

Comment:
style suggestions:
- instead of having `outgoing_socket` declared in the test function, and making `tester->socket *` point to it later, just have tester->socket be the actual client socket itself. It would just make the code simpler since there would be less things pointing at each other.
- also, could you rename this `client_socket`? I think that's more clear than `socket` or `outgoing_socket`